File: Task/Top-rank-per-group/C-sharp/top-rank-per-group.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Program
{
    class Employee
    {
        public Employee(string name, string id, int salary, string department)
        {
            Name = name;
            Id = id;
            Salary = salary;
            Department = department;
        }

        public string Name { get; private set; }
        public string Id { get; private set; }
        public int Salary { get; private set; }
        public string Department { get; private set; }

        public override string ToString()
        {
            return String.Format("{0, -25}\t{1}\t{2}", Name, Id, Salary);
        }
    }

    private static void Main(string[] args)
    {
        var employees = new List<Employee>
                        {
                            new Employee("Tyler Bennett", "E10297", 32000, "D101"),
                            new Employee("John Rappl", "E21437", 47000, "D050"),
                            new Employee("George Woltman", "E21437", 53500, "D101"),
                            new Employee("Adam Smith", "E21437", 18000, "D202"),
                            new Employee("Claire Buckman", "E39876", 27800, "D202"),
                            new Employee("David McClellan", "E04242", 41500, "D101"),
                            new Employee("Rich Holcomb", "E01234", 49500, "D202"),
                            new Employee("Nathan Adams", "E41298", 21900, "D050"),
                            new Employee("Richard Potter", "E43128", 15900, "D101"),
                            new Employee("David Motsinger", "E27002", 19250, "D202"),
                            new Employee("Tim Sampair", "E03033", 27000, "D101"),
                            new Employee("Kim Arlich", "E10001", 57000, "D190"),
                            new Employee("Timothy Grove", "E16398", 29900, "D190")
                        };

        DisplayTopNPerDepartment(employees, 2);
    }

    static void DisplayTopNPerDepartment(IEnumerable<Employee> employees, int n)
    {
        var topSalariesByDepartment =
            from employee in employees
            group employee by employee.Department
            into g
            select new
                    {
                        Department = g.Key,
                        TopEmployeesBySalary = g.OrderByDescending(e => e.Salary).Take(n)
                    };

        foreach (var x in topSalariesByDepartment)
        {
            Console.WriteLine("Department: " + x.Department);
            foreach (var employee in x.TopEmployeesBySalary)
                Console.WriteLine(employee);
            Console.WriteLine("----------------------------");
        }
    }
}

File: Task/Top-rank-per-group/Delphi/top-rank-per-group.pas
type TEmployeeInfo = record
 Name,ID: string;
 Salary: double;
 Dept: string;
 end;

var EmployeeInfo: array [0..12] of TEmployeeInfo = (
	(Name: 'Tyler Bennett'; ID: 'E10297'; Salary: 32000; Dept: 'D101'),
	(Name: 'John Rappl'; ID: 'E21437'; Salary: 47000; Dept: 'D050'),
	(Name: 'George Woltman'; ID: 'E00127'; Salary: 53500; Dept: 'D101'),
	(Name: 'Adam Smith'; ID: 'E63535'; Salary: 18000; Dept: 'D202'),
	(Name: 'Claire Buckman'; ID: 'E39876'; Salary: 27800; Dept: 'D202'),
	(Name: 'David McClellan'; ID: 'E04242'; Salary: 41500; Dept: 'D101'),
	(Name: 'Rich Holcomb'; ID: 'E01234'; Salary: 49500; Dept: 'D202'),
	(Name: 'Nathan Adams'; ID: 'E41298'; Salary: 21900; Dept: 'D050'),
	(Name: 'Richard Potter'; ID: 'E43128'; Salary: 15900; Dept: 'D101'),
	(Name: 'David Motsinger'; ID: 'E27002'; Salary: 19250; Dept: 'D202'),
	(Name: 'Tim Sampair'; ID: 'E03033'; Salary: 27000; Dept: 'D101'),
	(Name: 'Kim Arlich'; ID: 'E10001'; Salary: 57000; Dept: 'D190'),
	(Name: 'Timothy Grove'; ID: 'E16398'; Salary: 29900; Dept: 'D190')
	);


function SalarySort(Item1, Item2: Pointer): Integer;
var Info1,Info2: TEmployeeInfo;
begin
Info1:=TEmployeeInfo(Item1^);
Info2:=TEmployeeInfo(Item2^);
Result:=AnsiCompareStr(Info1.Dept,Info2.Dept);
If Result=0 then Result:=Trunc(Info1.Salary-Info2.Salary);
end;

procedure ShowTopSalaries(Memo: TMemo);
var List: TList;
var Info: TEmployeeInfo;
var I: integer;
var S,OldDept: string;

	procedure NewDepartment(Name: string);
	begin
	Memo.Lines.Add('');
	Memo.Lines.Add('Department: '+Name);
	Memo.Lines.Add('Employee Name    Employee ID      Salary           Department');
	OldDept:=Name;
	end;


begin
List:=TList.Create;
try
for I:=0 to High(EmployeeInfo) do
 List.Add(@EmployeeInfo[I]);
List.Sort(SalarySort);
OldDept:='';
for I:=0 to List.Count-1 do
	begin
	Info:=TEmployeeInfo(List[I]^);
	if OldDept<>Info.Dept then NewDepartment(Info.Dept);
	S:=Format('%-18S %9S %11.0m %8S',[Info.Name,Info.ID,Info.Salary,Info.Dept]);
	Memo.Lines.Add(S);
	end;

finally List.Free; end;
end;
